File: sample_app_dagger/src/main/java/com/chumarin/stanislav/sample_app_dagger/di/AppModule.java
package com.chumarin.stanislav.sample_app_dagger.di;

import android.content.Context;

import androidx.annotation.NonNull;

import com.chumarin.stanislav.sample_app_dagger.App;
import com.chumarin.stanislav.sample_app_dagger.Session;
import com.chumarin.stanislav.sample_app_dagger.SessionTrackerListener;

import java.util.Arrays;
import java.util.EnumSet;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;
import vit.khudenko.android.sessiontracker.ISessionStateTransitionsSupplier;
import vit.khudenko.android.sessiontracker.ISessionTrackerStorage;
import vit.khudenko.android.sessiontracker.SessionTracker;
import vit.khudenko.android.sessiontracker.Transition;

@Module
class AppModule {

    private static final String PREFS_FILENAME = "session_tracker_storage";

    @Provides
    @Singleton
    public SessionTracker<Session.Event, Session.State> provideSessionTracker(
            @NonNull ISessionTrackerStorage<Session.State> sessionStorage,
            @NonNull ISessionStateTransitionsSupplier<Session.Event, Session.State> stateTransitionsSupplier
    ) {
        return new SessionTracker<>(
                sessionStorage,
                stateTransitionsSupplier,
                EnumSet.of(Session.State.FORGOTTEN),
                SessionTracker.Mode.STRICT_VERBOSE,
                new SessionTracker.Logger.DefaultImpl(),
                "SessionTracker"
        );
    }

    @Provides
    @Singleton
    public ISessionStateTransitionsSupplier<Session.Event, Session.State> stateTransitionsSupplier() {
        return session -> Arrays.asList(
                new Transition<>(Session.Event.LOGIN, Arrays.asList(Session.State.INACTIVE, Session.State.ACTIVE)),
                new Transition<>(Session.Event.LOGOUT, Arrays.asList(Session.State.ACTIVE, Session.State.INACTIVE)),
                new Transition<>(Session.Event.LOGOUT_AND_FORGET, Arrays.asList(Session.State.ACTIVE, Session.State.FORGOTTEN))
        );
    }

    @Provides
    @Singleton
    public ISessionTrackerStorage<Session.State> sessionStorage(App app) {
        return new ISessionTrackerStorage.SharedPrefsImpl<>(
                app.getSharedPreferences(PREFS_FILENAME, Context.MODE_PRIVATE),
                EnumSet.allOf(Session.State.class)
        );
    }

    @Provides
    @Singleton
    public SessionTrackerListener sessionStateListener(App app) {
        return new SessionTrackerListener(app);
    }
}

File: sample_app_koin/src/main/kotlin/vit/khudenko/android/sessiontracker/sample/koin/di/appModule.kt
package vit.khudenko.android.sessiontracker.sample.koin.di

import android.content.Context
import org.koin.android.ext.android.getKoin
import org.koin.android.ext.koin.androidApplication
import org.koin.android.ext.koin.androidContext
import org.koin.core.qualifier.named
import org.koin.dsl.module
import vit.khudenko.android.sessiontracker.ISessionTrackerStorage
import vit.khudenko.android.sessiontracker.SessionTracker
import vit.khudenko.android.sessiontracker.Transition
import vit.khudenko.android.sessiontracker.sample.koin.Session.Event
import vit.khudenko.android.sessiontracker.sample.koin.Session.State
import vit.khudenko.android.sessiontracker.sample.koin.SessionTrackerListener
import java.util.EnumSet

val appModule = module {

    single(named(DI_NAME_SESSION_TRACKER_SHARED_PREFERENCES)) {
        androidContext().getSharedPreferences("session_tracker_storage", Context.MODE_PRIVATE)
    }

    single {
        SessionTracker(
            sessionTrackerStorage = ISessionTrackerStorage.SharedPrefsImpl<State>(
                get(named(DI_NAME_SESSION_TRACKER_SHARED_PREFERENCES)),
                EnumSet.allOf(State::class.java)
            ),
            sessionStateTransitionsSupplier = {
                listOf(
                    Transition(event = Event.LOGIN, statePath = listOf(State.INACTIVE, State.ACTIVE)),
                    Transition(event = Event.LOGOUT, statePath = listOf(State.ACTIVE, State.INACTIVE)),
                    Transition(event = Event.LOGOUT_AND_FORGET, statePath = listOf(State.ACTIVE, State.FORGOTTEN))
                )
            },
            autoUntrackStates = setOf(State.FORGOTTEN),
            mode = SessionTracker.Mode.STRICT_VERBOSE
        )
    }

    single {
        SessionTrackerListener { androidApplication().getKoin() }
    }

}
